### PR TITLE
feat(lora): add FIM-guided automatic LoRA rank allocation (fim_auto_rank)

### LINKED ISSUE #3626: [Feature] Automatic LoRA rank recommendation based on dataset size
## Description

Choosing the right LoRA rank is largely trial-and-error. An automatic recommendation based on dataset size, task type, and base model would save time.

## Use Case

- New users don't know what rank to choose
- Different tasks benefit from different ranks (classification vs generation)
- Could be a `--auto-rank` flag or a preprocessing analysis step

Based on our experience, rank 16 works for most tasks under 10K samples, but this heuristic could be formalized.

### PR DESCRIPTION
**Pull request by ramkrishs**
## Summary

Adds `fim_auto_rank: true` config option that automatically selects per-layer LoRA ranks using the diagonal of the empirical Fisher Information Matrix (eFIM) — concentrating rank budget on layers with the highest loss sensitivity and reducing rank on less sensitive layers.

Closes #3626

---

## Motivation

Choosing the right LoRA rank is trial-and-error. The FIM diagonal provides a principled, data-driven answer: layers with high gradient variance are more sensitive to the loss and benefit from higher rank; insensitive layers can use lower rank without hurting performance.

**Algorithm:**
```
F_ii ≈ (1/T) Σ (∂ℓ_t / ∂θ_i)²    # eFIM diagonal (mean squared gradient)
score_i = mean(F_i)                 # per-layer importance
rank_i ∝ score_i / Σ score_j × budget   # budget = n_layers × lora_r
rank_i = clamp(rank_i, fim_rank_min, fim_rank_max)
```

The mean rank across layers is preserved — total parameter budget is unchanged.

---

## Usage

```yaml
adapter: lora
lora_r: 16
fim_auto_rank: true
fim_calibration_batches: 8   # default 8
fim_rank_min: 1               # default 1
fim_rank_max: 32              # default 2 × lora_r
```

---

## Files changed

| File | Change |
|------|--------|
| `src/axolotl/utils/fim_rank.py` | New: FIM accumulation, rank allocation, adapter resizing |
| `src/axolotl/utils/schemas/peft.py` | 4 new config fields |
| `src/axolotl/loaders/adapter.py` | Calls `apply_fim_ranks()` after `get_peft_model()` |
| `tests/utils/lora/test_fim_rank.py` | 13 unit tests (all passing) |

---

## Reference

- LeCun et al., *Optimal Brain Damage*, NeurIPS 1990
- Related: [huggingface/peft#3204](https://github.com/huggingface/peft/pull/3204) — same eFIM approach as a PEFT-native `init_lora_weights='fim'` mode

> This PR was developed with AI assistance. All code has been reviewed, tested, and verified to follow axolotl conventions.

## Release Notes

* **New Features**
  * Added Fisher Information Matrix-guided automatic LoRA rank allocation to optimize layer ranks based on empirical importance during calibration.
  * New configuration options: `fim_auto_rank`, `fim_calibration_batches`, `fim_rank_min`, and `fim_rank_max` for controlling rank allocation behavior.

* **Tests**
  * Added comprehensive test suite validating FIM rank allocation functionality and constraints.

### CODE REVIEW

**Comment by ramkrishs** [comment]
Addressing the CodeRabbit pre-merge check comments:

**Docstring coverage** — fixed in the latest commit. Added full Google-style `Args`/`Returns` docstrings to all functions in `fim_rank.py`. All 13 tests still pass.

**Linked Issues check (eFIM vs dataset-size heuristics)** — the original issue (#3626) describes the *problem* as trial-and-error rank selection and suggests dataset-size heuristics as one possible solution. The eFIM approach directly solves the underlying problem in a more principled way:

- Dataset-size heuristics (e.g. "rank 16 for <10K samples") are model-agnostic rules of thumb that ignore actual layer sensitivity.
- eFIM-based allocation measures **per-layer loss sensitivity** directly from the data+model combination at hand — it automatically adapts to the task, base model architecture, and training distribution.
- The calibration pass uses the same training data, so no extra data collection is needed.

In practice this means a user who sets `fim_auto_rank: true` gets an automatically tuned rank distribution without needing to know anything about LoRA theory — which is the core use case in #3626. The implementation also falls back gracefully (warning + unchanged ranks) if no calibration dataloader is available, so it never breaks existing workflows.

**Comment by NanoCode012** [comment]
@ramkrishs , thanks for the PR. It looks like you also pushed to peft upstream. Similar to the peft maintainers, we would be happy to pick up once your preprint is ready.

However, this seems more like something that should be merged upstream?

**Comment by ved1beta** [comment]
stale closing
